Move the Foreman Blog to Markdown/Jekyll/GitHub

GregSutcliffe opened the pull request:
So this is fairly big, but the change is fairly smooth. Overall, the content looks pretty good. Some notes:
- Content imported via Pandoc, which made a pretty good job of the conversion to Markdown
- Images moved to be local to the blog so we don't rely on Blogger at all
- Gists correctly restored
- RSS feed add
- Blog search added via Google site search box

The search, G+/Twitter buttons, and Comments are untested, because they rely on the permalink, which is (of course) incorrect when running on localhost:4000. Should be easy enough to test and roll forward once deployed.

Regarding the permalink, the blog URLs should be correct, on the basis that we CNAME blog.theforeman.org to theforeman.org. If that doesn't work, we'll need to investigate potential fixes.

Following the discussion on #446 about blog quality, I've also added a separate commit which re-organises the Community page a little, and now adds the blog to the main navbar, since it's quality should improve from here. I'll be happy to move that to a separate PR if desired.

I think this is ready to merge, but we need the DNS change on standby for when it's actually deployed. I also have a new blog post (not in the PR, but written and waiting) announcing the revamped blog. I'll publish that to both the new and old blogs once this is deployed.

GregSutcliffe committed:
Initial Markdown blog support

Google Plus comments (i.e Blogger comments) work off the permalink,
so we match it. This means blogs are at /year/month while the index is
at /blog/, but thats a minor hiccup.

Adds an RSS feed at /feed.xml

Also adds initial search, style, and social sharing buttons. This
will need improving.

Commit:
Updating blog style

- search field positioned to right
- search and back button have the same style
- positioning of back button and social buttons
- well around the sharing encouragement text
- better visual structure for post headings
- blog pagination uses standard bootstrap pager

GregSutcliffe committed:
Import posts from Blogger, convert with Pandoc; fix review comments

Also includes downloading the images from Blogger and using
the 'gist' MD directive from the jekyll-gist gem to render code
snippets

GregSutcliffe committed:
Switch to jekyll-feed; Add RSS icon and text to blog index

GregSutcliffe commented:
The more I think about it, the more I think the navbar changes should be a new PR. I've removed that commit for now.

domcleal commented on _includes/header.html:
```diff
@@ -12,6 +12,7 @@
     {% assign pagename = true %}
     {% endif %}
   <title>{{ sitename }}{% if pagename %} :: {{ page.pagename }}{% endif %}</title>
+  <link rel="alternate" type="application/rss+xml" title="RSS" href="http://{{ site.url }}/feed.xml" />
```
Just use {{site.baseurl}} rather than protocol + url?

GregSutcliffe replied:
Good catch, fixed.

domcleal commented on Rakefile:
```diff
@@ -19,3 +19,23 @@ desc 'Starts up browser window (requires running server)'
   sh "xdg-open http://localhost:4000/index.html"
 end
+
+desc "Given a title as an argument, create a new post file"
```
Can you put this in README.md?

GregSutcliffe replied:
Yep, done.

domcleal commented on Rakefile:
```diff
@@ -19,3 +19,23 @@ desc 'Starts up browser window (requires running server)'
+layout: post
+title: #{args.title}
+date: #{Time.now.strftime('%Y-%m-%d %k:%M:%S')}
+author: Foreman
```
Perhaps use Etc.getpwnam etc?

GregSutcliffe replied:
To make a guess at the current author? I'd expect that to nearly always be wrong - on my system:

```
Etc.getpwnam(Etc.getlogin).name
=> "greg"
```

I think we're just overcomplicating the task when it's something people will need to edit anyway (and they'll have to edit the metadata anyway for tags and such). I've added a "check the author" step to the README, which I think is sufficient - will that do for you?

domcleal replied:
.gecos, not .name.  Adding a step's fine too, I just figured it could work the same way as git.

GregSutcliffe replied:
Heh, .gecos is "" on my laptop because I'm lazy. See if the README changes work for you :)

domcleal commented on feed.xml:
```diff
@@ -0,0 +1,27 @@
+---
```
We have jekyll-feed available, can it do this?  https://github.com/jekyll/jekyll-feed

GregSutcliffe replied:
Can do. That's Atom rather than RSS2.0 - do we care?

domcleal replied:
I don't think so, Atom's supported just as much nowadays - it's equivalent.

GregSutcliffe replied:
Fair enough. Seemed like jekyll-feed had the same discussion earlier this year and it was an involved discussion, so I thought I might ask ;)

I'll try and get that implemented now.

GregSutcliffe replied:
Works very smoothly by default. I've added the auto-discovery helper but I'm not sure if it's working for me. Can someone else confirm?

domcleal replied:
Working fine here, I see the meta tag.

domcleal commented on _layouts/post.html:
```diff
@@ -0,0 +1,56 @@
+      </div>
+
+      <div class="well">
+        You can follow us on Twitter: @ForemanProject. If you tweet about foreman, please use the hashtag #theforeman so we can see it!<br />
```
about "Foreman"

Missing hyperlinks?

GregSutcliffe replied:
Fixed, thanks.

domcleal commented on _includes/search.html:
```diff
@@ -0,0 +1,25 @@
+<div class="container container-search">
+  <a class="pull-left" href="/feed.xml">
+    <img src="/static/images/rss.png" alt="RSS Feed" />
+    Get the RSS News Feed
```
Not sure this caption's needed?

GregSutcliffe replied:
I'm happy to change the wording, but the icon alone is easily missed (especially as it's not visible anywhere else), and with the search bar to the right we have the space for _something_. Suggestions welcome for improvements welcome - I wasn't 100% happy with it either.

domcleal replied:
Actually, thinking about it again, I'm OK with it staying if we can get a proper heading above it.  What bothered me was that it seemed to be the page heading, so it distracted you from reading the page content.

What do you think of https://github.com/domcleal/theforeman.org/commit/f0358390aebc34cc97035998cc818b43d43dde70?  It reduces the font sizes again, but adds in a page heading which helps refocus the page IMHO.

GregSutcliffe replied:
I like it more, certainly. There's some wasted whitespace above the searchbox now, but I think I'm ok with that. I guess the final option would be to put the RSS link and the search together on the right (one over the other) opposite the heading.

Added your commit to the PR as it's definitely an improvement. Thanks!

domcleal commented on blog/index.html:
```diff
@@ -0,0 +1,44 @@
+
+{% for post in paginator.posts %}
+  <div class="blog-item">
+    <h1><a href="{{ post.url }}">{{ post.title }}</a></h1>
```
Very, very big.  Try h2?

GregSutcliffe replied:
Done, thanks.

domcleal commented:
1. There are no links to the blog that I can see!  Do change the link on the documentation index from blog.tf.org.
2. The search toolbar isn't working for me, the page is simply refreshing.
3. The text size is 18pt compared to our usual 14pt for docs, introduction pages etc.  Can you apply a 14pt size (look at the `doc` class) and then remove the bold headings, as they'll no longer be required?  It should look a lot more similar to the manual then.

GregSutcliffe commented:
@domcleal re: (2) Initially I thought it was because it's running locally so I didn't pay too much mind to it. Now I see it's just a Google site-specific search in a new window, if I'm reading JS right (@tstrachota added it). My JS is awful, as is well known, so I'm not sure how to fix it - any ideas?

GregSutcliffe commented:
Re: (3) I've just updated the post and post_index content id to use the `doc` class so we're always consistent if it changes. Seems to look good here - thanks!

EDIT: applying the same to blog/index.html is causing problems though - it makes the excerpt space _huge_ even when it's empty. Not sure why - any ideas...?

GregSutcliffe commented:
(1) - fixed, thanks.

GregSutcliffe commented:
@domcleal fixes so far pushed, thanks. A few open questions for you ... :)

domcleal commented on README.md:
```diff
@@ -29,6 +29,15 @@ To start Jekyll server locally.
+## Contributing a blog post
+
+1. (following on from step 5 above)
+2. Run rake new\_post['title of my post'] to generate a new empty blog post
```
Using backtick (`) would be easier to read

GregSutcliffe replied:
D'oh. Fixed.

domcleal commented on Rakefile:
```diff
@@ -19,3 +19,23 @@ desc 'Starts up browser window (requires running server)'
+tags:
+- foreman
+---
+EOS
```
Should this include the "more" comment?

GregSutcliffe replied:
Yeah, good catch, I'll flesh that out

domcleal commented on blog/index.html:
```diff
@@ -0,0 +1,44 @@
+    <p class="metadata">
+      <span class="author">{{ post.author }}</span> - <span class="date">{{ post.date | date: "%-d %B %Y" }}</span>
+    </p>
+    <div class="bold-headings">
```
No need for bold-headings?

GregSutcliffe replied:
Done.

domcleal commented on static/css/style.css:
```diff
@@ -1062,3 +1073,31 @@ code {
+  padding-bottom: 20px;
+}
+
+.bold-headings h1,
```
unused?

GregSutcliffe replied:
Done.

domcleal commented on static/css/style.css:
```diff
@@ -1062,3 +1073,31 @@ code {
+}
+
+.post-content {
+  padding-top: 20px;
```
I'd suggest removing this, there's a large gap between the blog post title and the content.

domcleal commented on _layouts/post.html:
```diff
@@ -0,0 +1,56 @@
+
+      <hr/>
+
+      <div class="post-content" id="doc">
```
One minor downside about using a `doc` element is that it sets a minimum height of 500px, might be better to just have the attribute(s) we need for blogs?

GregSutcliffe replied:
Ah, that would explain my issue. I'll copy the relevant parts of #doc to .post-content then. Thanks!

domcleal commented on _layouts/post_index.html:
```diff
@@ -0,0 +1,9 @@
+{% include search.html %}
+
+<div id="wrap">
+	<div class="container" id="doc">
```
Change tabs to spaces?

domcleal commented on _layouts/post.html:
```diff
@@ -0,0 +1,56 @@
+      </div>
+
+      <div class="well">
+        You can follow us on Twitter: <a href='http://twitter.com/ForemanProject'>@ForemanProject</a>. If you tweet about Foreman, please use the hashtag #theforeman so we can see it!<br />
```
Can you reduce the font size here too?  It's also 18px, larger than the blog post itself.

GregSutcliffe replied:
Done.

Commit:
Add a blog index heading, reduce font sizes

domcleal commented:
> EDIT: applying the same to blog/index.html is causing problems though - it makes the excerpt space huge even when it's empty. Not sure why - any ideas...?

Ah, this is what I mentioned in an inline comment just now - there's a min size on #doc, so I'd suggest creating a class specifically for the blog.  (#doc is also no use on the index since it's an ID and not a class.)

domcleal commented:
If you get the opportunity, I think it would be really useful to include an excerpt for the first five posts so the index page looks a bit more fleshed out to begin with.

domcleal commented:
Some of the CSS changes seem to have changed the appearance of the buttons on the homepage too.

Before:
![screenshot from 2015-11-02 11-11-52](https://cloud.githubusercontent.com/assets/482089/10880703/61ca82e8-8155-11e5-806a-79402b2f0b7f.png)

After:
![screenshot from 2015-11-02 11-11-59](https://cloud.githubusercontent.com/assets/482089/10880702/60aa6e50-8155-11e5-8189-cb9bac2068e1.png)

domcleal commented on _includes/search.html:
```diff
@@ -0,0 +1,25 @@
+<script language="Javascript" type="text/javascript">
+  function my_search_google()
+  {
+    var query = document.getElementById("my-google-search").value;
```
The ID is just `search` for the text box.

domcleal commented on _includes/search.html:
```diff
@@ -0,0 +1,25 @@
+      <div class="input-group">
+        <input type="text" class="form-control" id="search" placeholder="Search">
+        <span class="input-group-btn">
+          <button type="submit" class="btn btn-foreman-blue" onsubmit="my_search_google()">Search</button>
```
onsubmit needs to be on the form element

GregSutcliffe replied:
Got it working here, thanks for the pointers!

domcleal commented:
> My JS is awful, as is well known, so I'm not sure how to fix it - any ideas?

Comments inline.

domcleal commented:
I get a large spinner beneath the comments box:

![screenshot from 2015-11-02 11-56-29](https://cloud.githubusercontent.com/assets/482089/10881207/d77fd404-8158-11e5-805a-bdb281a3b095.png)

Is that related to this?

> The search, G+/Twitter buttons, and Comments are untested, because they rely on the permalink, which is (of course) incorrect when running on localhost:4000. Should be easy enough to test and roll forward once deployed.

domcleal commented on _layouts/post.html:
```diff
@@ -0,0 +1,56 @@
+
+      <div class="row">
+        <div class="col-md-6 col-sm-6 col-xs-6">
+          <a href="{{site.baseurl}}blog/" class="btn btn-foreman-blue"><span aria-hidden="true">&larr;</span> Return</a>
```
Nitpick, this has a different style to the prev/next buttons on the blog index - both in the button size/shape/colour and capitalisation.

GregSutcliffe replied:
I'm struggling to fix this to be honest - even adding the `btn btn-foreman-blue` classes isn't fixing the shape, and the colours are being overridden somewhere else. If it's just a nitpick can we record it and come back to it?

domcleal replied:
Sure.

GregSutcliffe replied:
#466 recorded for that, thanks.

GregSutcliffe commented:
@domcleal re the spinner, yeah, it's because the URL is localhost, I think. Really hard to test that in advance, but the fact that the comment box appears at all is a good sign. At least we're loading it correctly.

Re the frontpage buttons, it's an accident, but I like the new ones. Are you happy with it too?

Comments addressed I think. I'll sort out the nitpick and the excepts now, I think you make a good point there.

domcleal commented:
> Re the frontpage buttons, it's an accident, but I like the new ones. Are you happy with it too?

Yep, fine by me.

domcleal commented on _posts/2015-09-15-foreman-community-newsletter-september.md:
```diff
@@ -0,0 +1,109 @@
+blogger_orig_url: http://blog.theforeman.org/2015/09/foreman-community-newsletter-september.html
+---
+
+### Foreman 1.9 and 1.9.1 released
```
This heading appears on the index page inside the excerpt and sort of ruins the layout.  Could you include an excerpt manually like October's?

GregSutcliffe replied:
Well, you could level that at all of them, so to pre-empt you, I've gone ahead and added an excerpt to all of them :)

I do think the below-the-fold style will useful from time-to-time though. I'll keep it in mind when writing/reviewing posts.

domcleal replied:
Ah thanks.  Yes, I agree it will and it ultimately depends on the post content.

On another site I maintain I swap between both as sometimes if you're putting images or something into the text then you need to set a manual excerpt.

domcleal commented on _posts/2015-10-07-foreman-community-newsletter-october.md:
```diff
@@ -0,0 +1,65 @@
+found in the
+[announcement](https://groups.google.com/forum/#!topic/foreman-announce/TQqnzHmBJOY).
+
+### **Foreman 1.8.4 released**
```
Why are these bold (now)?

GregSutcliffe replied:
I don't think I added that - I did _remove_ bold from the 1.9.2 header while attempting to sort out the excerpts, which is possibly why they now look inconsistent. I'll put it back.

domcleal replied:
Please just remove it, don't add it back!

GregSutcliffe replied:
I've removed the bold from all the headings, to be consistent with the other newsletters. So much for not changing the author's work - that's how @orrabin styled it ;)

domcleal commented on static/css/style.css:
```diff
@@ -1062,3 +1073,35 @@ code {
+}
+
+.post-content h5 {
+	font-size: 16px;
```
Change tabs to spaces, please.

GregSutcliffe replied:
Grr, I have no idea how that's happening. Will fix, thanks.

domcleal replied:
Still here.

GregSutcliffe replied:
Fixed, thanks.

domcleal commented on static/css/style.css:
```diff
@@ -849,18 +851,27 @@ fieldset[disabled] .source-buttons .btn-primary.active {
-  letter-spacing: .15em;
+  letter-spacing: .05em;
   font-weight: bold;
-  font-size: .9em;
```
I think this is also changing the font size on the documentation page - see the difference in size of the text underneath each doc icon.

domcleal replied:
And probably also contributing to the homepage buttons!

GregSutcliffe replied:
Yeah, it does have an effect, although they are still different. I've reverted it, since I'm not sure why @tstrachota did that. There's a few other changes in that class, do you want them all reverting?

domcleal replied:
Yeah, the layout of the docs page still isn't the same.  I think you need to only apply these changes to where you're using them, which I think is only the "Return" button on the post page.

domcleal commented on _posts/2015-04-30-foreman-community-newsletter-april-2015.md:
```diff
@@ -0,0 +1,115 @@
+blogger_orig_url: http://blog.theforeman.org/2015/04/foreman-community-newsletter-april-2015.html
+---
+
+<span style="background-color: white; color: black; font-family: Arial; font-size: 20px; font-style: normal; font-variant: normal; font-weight: normal; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Foreman Community Newsletter - </span><span style="background-color: transparent; color: black; font-family: 'Trebuchet MS'; font-size: 21px; font-style: normal; font-variant ...
```
While I don't want to completely rewrite the blog posts, this one's a formatting mess...

domcleal commented on _posts/2015-01-30-foreman-community-newsletter-january.md:
```diff
@@ -0,0 +1,87 @@
+
+<!--more-->
+
+### <span style="background-color: white; color: #333333; font-family: Arial; font-size: 17px; font-style: normal; font-variant: normal; font-weight: bold; text-decoration: none; vertical-align: baseline; white-space: pre-wrap;">Foreman 1.7.2 released</span>
```
This one too - perhaps we can just filter out these sorts of stylings?

GregSutcliffe replied:
I've already run it through pandoc once, I think anything further would have to be by hand, annoyingly. It's not on the first page of results, and I doubt the newsletters get high hits after a couple of months. Can we add this to the nitpick list?

domcleal replied:
Okay.

GregSutcliffe replied:
#467 added for that, thanks.

domcleal commented:
To add to the inline comment about the changes to the btn class, I think removing some of the customisations entirely has broken the search button too:

![screenshot from 2015-11-02 15-24-13](https://cloud.githubusercontent.com/assets/482089/10885596/edcd6902-8175-11e5-809e-c413906abfc3.png)

Separating the customisations from the regular class ought to fix the existing and new uses.

GregSutcliffe commented:
I've reverted @tstrachota's changes to the `btn` class and added the relevant parts to the `btn-foreman-blue` class instead, so that should drop all the unintended side effects on the the rest of the site. Looks ok here. I can open a separate tiny PR to add the rounded corners back to the main `btn` class once we're done with this.

domcleal commented:
Okay, that looks better.  Will you just add a post to the old site to notify users, or can you add a redirect to it somehow?

GregSutcliffe commented:
I have a blog post about the new blog ready to go - I figured I'd post it to both sites once we're ready to switch. I can add it to this PR, if you like. I'll look into a redirect as well, good thought.

domcleal commented:
Yeah, perhaps if the template/layout of the old blog can be edited (I assume so, to add analytics etc) then a small piece of JavaScript could redirect to the new URL at least.

There was one inline comment noted above.

GregSutcliffe commented:
Indeed, looks possible - eg http://smallbusiness.chron.com/make-blogspot-redirect-another-69368.html

GregSutcliffe commented:
So I think it makes sense to do belt'n'braces - add a post to the old site, and then once we're happy that the new site is working correctly (social buttons, comments etc) we can add the auto redirect.

domcleal commented:
Thanks @GregSutcliffe, merged.
